### PR TITLE
Fix TableJoin to require distinct field names

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala [hunk 1]
```diff
@@ -239,18 +239,18 @@ object PruneDeadFields {
         val leftDep = left.typ.copy(
           rowType = TStruct(left.typ.rowType.required, left.typ.rowType.fieldNames.flatMap(f =>
             requestedType.rowType.fieldOption(f).map(reqF => f -> reqF.typ)): _*),
-          globalType = requestedType.globalType)
+          globalType = TStruct(left.typ.globalType.required, left.typ.globalType.fieldNames.flatMap(f =>
+            requestedType.globalType.fieldOption(f).map(reqF => f -> reqF.typ)): _*))
         memoizeTableIR(left, leftDep, memo)
-
-        val rightKeyFields = right.typ.key.iterator.flatten.toSet
+        val rightKeyFields = right.typ.key.getOrElse(FastIndexedSeq()).toSet
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala]: `right.typ.keyOrEmpty.toSet` is shorter, and easier to clean up when `key` (soon) becomes required on `TableType`.

File: hail/src/main/scala/is/hail/expr/ir/TableIR.scala [hunk 2]
```diff
@@ -396,8 +396,14 @@ case class TableJoin(left: TableIR, right: TableIR, joinType: String, joinKey: I
   require(joinKey >= 0)
   require(left.typ.keyType.zip(right.typ.keyType).exists { case (leftKey, rightKey) =>
     (leftKey.size >= joinKey) && (rightKey.size >= joinKey) &&
-      (leftKey.truncate(joinKey) isIsomorphicTo rightKey.truncate(joinKey))
+    (leftKey.truncate(joinKey) isIsomorphicTo rightKey.truncate(joinKey))
   })
+  require(left.typ.valueType.fieldNames.toSet
+    .intersect(right.typ.valueType.fieldNames.toSet)
+    .isEmpty)
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/TableIR.scala]: I think this should be using `leftRVDType` and `rightRVDType` below. Key fields that aren't part of the join key shouldn't collide either.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/TableIR.scala]: that's a good point, yes

**Comment by tpoterba** [comment]
fixes #4311

**Pull request merged**